SU2_CFD/src/solvers/CEulerSolver.cpp

```diff
@@ -3426,238 +3122,226 @@ void CEulerSolver::Upwind_Residual(CGeometry *geometry, CSolver **solver_contain
-
-      numerics->SetPrimitive(V_i, V_j);
-      numerics->SetSecondary(S_i, S_j);
+      numerics->SetPrimitive(bad_i? V_i : Primitive_i,  bad_j? V_j : Primitive_j);
+      numerics->SetSecondary(bad_i? S_i : Secondary_i,  bad_j? S_j : Secondary_j);
 
     }
 
     /*--- Roe Low Dissipation Scheme ---*/
 
-    if (kind_dissipation != NO_ROELOWDISS){
+    if (kind_dissipation != NO_ROELOWDISS) {
 
-      Dissipation_i = nodes->GetRoe_Dissipation(iPoint);
-      Dissipation_j = nodes->GetRoe_Dissipation(jPoint);
-      numerics->SetDissipation(Dissipation_i, Dissipation_j);
+      numerics->SetDissipation(nodes->GetRoe_Dissipation(iPoint),
+                               nodes->GetRoe_Dissipation(jPoint));
 
       if (kind_dissipation == FD_DUCROS || kind_dissipation == NTS_DUCROS){
-        Sensor_i = nodes->GetSensor(iPoint);
-        Sensor_j = nodes->GetSensor(jPoint);
-        numerics->SetSensor(Sensor_i, Sensor_j);
+        numerics->SetSensor(nodes->GetSensor(iPoint),
+                            nodes->GetSensor(jPoint));
       }
       if (kind_dissipation == NTS || kind_dissipation == NTS_DUCROS){
-        Coord_i = geometry->node[iPoint]->GetCoord();
-        Coord_j = geometry->node[jPoint]->GetCoord();
         numerics->SetCoord(Coord_i, Coord_j);
       }
     }
 
     /*--- Compute the residual ---*/
 
-    numerics->ComputeResidual(Res_Conv, Jacobian_i, Jacobian_j, config);
+    auto residual = numerics->ComputeResidual(config);
 
-    /*--- Update residual value ---*/
+    /*--- Set the final value of the Roe dissipation coefficient ---*/
 
-    LinSysRes.AddBlock(iPoint, Res_Conv);
-    LinSysRes.SubtractBlock(jPoint, Res_Conv);
+    if ((kind_dissipation != NO_ROELOWDISS) && (MGLevel != MESH_0)) {
+      nodes->SetRoe_Dissipation(iPoint,numerics->GetDissipation());
+      nodes->SetRoe_Dissipation(jPoint,numerics->GetDissipation());
+    }
 
-    /*--- Set implicit Jacobians ---*/
+    /*--- Update residual value ---*/
 
-    if (implicit) {
-      Jacobian.UpdateBlocks(iEdge, iPoint, jPoint, Jacobian_i, Jacobian_j);
+    if (reducer_strategy) {
+      EdgeFluxes.SetBlock(iEdge, residual);
+      if (implicit)
+        Jacobian.UpdateBlocks(iEdge, residual.jacobian_i, residual.jacobian_j);
     }
+    else {
+      LinSysRes.AddBlock(iPoint, residual);
+      LinSysRes.SubtractBlock(jPoint, residual);
 
-    /*--- Set the final value of the Roe dissipation coefficient ---*/
-
-    if (kind_dissipation != NO_ROELOWDISS){
-      nodes->SetRoe_Dissipation(iPoint,numerics->GetDissipation());
-      nodes->SetRoe_Dissipation(jPoint,numerics->GetDissipation());
+      /*--- Set implicit computation ---*/
+      if (implicit)
+        Jacobian.UpdateBlocks(iEdge, iPoint, jPoint, residual.jacobian_i, residual.jacobian_j);
     }
 
+    /*--- Viscous contribution. ---*/
+
+    Viscous_Residual(iEdge, geometry, solver_container,
+                     numerics_container[VISC_TERM + omp_get_thread_num()*MAX_TERMS], config);
```

Comment:
For a bit of extra performance and generality in the context of that scatter-to-gather transformation, I merged the convective and viscous loops, I've done this the polymorphic way with a `Viscous_Residual` method that acts on a single edge and only the NS solver implements.

I may yet do the same for the turbulence solvers since there it makes less sense to have separate convective and viscous loops.